chat_client/src/main.py

```diff
@@ -4,119 +4,186 @@
-    run():
-        Run chat client
-    disconnect():
-        Close channel, receiver thread and any other stuff opened.
-        Must be called at the end of object existance.
-    """
+    """A class to represent a chat client object."""
 
     def __init__(self, host: str, port: int) -> None:
         """Constructs all the necessary attributes for the client object.
 
-        Parameters
-        ----------
-            host : str
-                Host address(Address of the server)
-            port : int
-                Port number
+        Args:
+            host (str): Host address(Address of the server).
+            port (int): Port number.
         """
         self._is_connected = False
 
         self._channel = None
         self._stub = None
         self._receiver = None
 
-        self._login = ""
+        self._username = ""
         self._connection_addr = f"{host}:{port}"
         logging.debug("Chat client object created")
 
     def connect(self) -> None:
-        """Ask for username and connect."""
+        """Asks for username and connect."""
         if self._is_connected:
             return
-        if not self._login:
-            self._set_login_name()
         try:
             self._channel = grpc.insecure_channel(self._connection_addr)
             self._stub = chat_pb2_grpc.ChatServiceStub(self._channel)
-        except Exception as e:
+            self._handle_register()
+            self._handle_login()
+        except ConnectionRefusedError as e:
             logging.error("Cannot connect [%s]", e)
+            self.disconnect()
         else:
             self._is_connected = True
             logging.info("Chat client connected")
 
-    def change_login_name(self) -> None:
-        """Change or set user login name"""
-        self._set_login_name()
+    def _handle_register(self) -> None:
+        """Handles user register, takes username, full name and password from user.
+        User have 3 chances to input valid password.
+
+        Raises:
+            rpc_error: Raised when error type was not expected.
+        """
+        if input(
+            "Do you want to register first? yes/[no]"
+        ).strip().lower() in ["y", "yes"]:
+            username = self._get_username()
+            full_name = input("Full name:").strip()
+            password = getpass()
+            req = chat_pb2.RegisterUserRequest(
+                user_info=chat_pb2.UserInfo(
+                    login=username, full_name=full_name
+                ),
+                password=password,
+            )
+            try:
+                self._stub.RegisterUser(request=req)
+            except grpc.RpcError as rpc_error:
+                if rpc_error.code() == grpc.StatusCode.ALREADY_EXISTS:
+                    logging.info("User %s already exists...", username)
+                elif rpc_error.code() == grpc.StatusCode.UNAVAILABLE:
+                    logging.debug("Server unavaible...")
+                else:
```

Comment:
Try to not use else without certain need, leave it as is, in your next task I'll show you why it's not a good practice;)

Reply:
Sure, so there is no changes for now